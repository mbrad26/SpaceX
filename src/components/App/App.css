body {
  font-family: 'Space Mono', monospace;
  /* background-color: black; */
  background-color: #e5e5e5;
  /* background: url(./falcon-9-rocket-4928x3264-spacex-cape-canaveral-4k-6186.jpg) no-repeat center center fixed;
  -webkit-background-size: cover;
  -moz-background-size: cover;
  -o-background-size: cover;
  background-size: cover; */
}

#title {
  padding: 20px;
  width: 1000px;
}

.container {
  padding: 20px;
  width: 850px;
}

/* Navbar */

#navbar {
  z-index: 1;
}

nav {
  margin: 20px 75px;
}

.links-container {
  padding: 0px;
  margin: 0px;
}

/*
img {
  display: block;
  width: 100%;
  height: auto;
} */

btn {
  border-style: hidden;
  border: none;
}

/* List and Item */

.row {
  padding: 20px;
}

.card-title {
  font-size: 1.55rem;
  /* font-weight: normal; */
}

.item-description {
  font-size: 0.9rem;
  overflow: hidden;
  display: -webkit-box;
  -webkit-line-clamp: 5;
  -webkit-box-orient: vertical;
}

.card-img-top {
  width: 100%;
  height: 35vh;
}

/* Modal */

Modal .container-fluid {
  padding: 16px;
}

#wiki {
  margin-bottom: 0px;
}

.row {
  padding: 0px 0px;

}

.col {
  padding:  0px;
  font-size: 0.9rem;
  justify-content: center;
}

.carousel-item {
  width: 100%;
  height: 375px;
  overflow: hidden;
}
.carousel-item img {
  width: 100%;
  height: 375px;
  object-fit: cover;
}
