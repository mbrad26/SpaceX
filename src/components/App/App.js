import React, {
  useEffect,
  useReducer,
  useCallback,
  useRef,
} from 'react';
import List from '../List/List';
import NavBar from '../NavBar/NavBar';
import ModalComponent from '../Modal/Modal';
import Container from "react-bootstrap/Container";
import Row from "react-bootstrap/Row";
import Col from "react-bootstrap/Col";
import axios from 'axios';
import './App.css';

const Context = React.createContext(null);

const API_ENDPOINT = 'https://api.spacexdata.com/v3/';

const dataReducer = (state, action) => {
  switch (action.type) {
    case 'DATA_LOADING':
      return {
        ...state,
        loading: true,
        isError: false,
      };
    case 'DATA_SUCCESS':
      return {
        ...state,
        loading: false,
        data: action.payload,
      };
    case 'DATA_ERROR':
      return {
        ...state,
        loading: false,
        isError: true,
      }
    case 'SET_URL':
      return {
        ...state,
        url: `${API_ENDPOINT}${action.payload}`,
      }
    case 'OPEN_MODAL':
      return {
        ...state,
        isOpen: true,
        activeItem: action.payload,
      }
    case 'CLOSE_MODAL':
      return {
        ...state,
        isOpen: false,
      }
    default:
      throw new Error();
  }
};

const App = () => {
  const initialState = {
    data: [],
    loading: false,
    isError: false,
    url: '',
    isOpen: false,
    activeItem: '',
  };
  const [state, dispatch] = useReducer(dataReducer, initialState);
  const isMounted = useRef(false);

  const fetchData = useCallback(async () => {
    dispatch({ type: 'DATA_LOADING' });

    try {
      const result = await axios.get(state.url);
      dispatch({
        type: 'DATA_SUCCESS',
        payload: result.data
      });
    } catch {
        dispatch({ type: 'DATA_ERROR' });
    }
  }, [state.url]);

  const handleClick = event => {
    dispatch({ type: 'SET_URL', payload: event.target.innerHTML });
  }

  const handleOpenModal = item => {
    dispatch({ type: 'OPEN_MODAL', payload: item});
  }

  const handleCloseModal = () => {
    dispatch({ type: 'CLOSE_MODAL' });
  }

  useEffect(() => {
    if(!isMounted.current) {
      isMounted.current = true;
    } else {
      fetchData();
    }
  }, [fetchData]);

  return (
    <Context.Provider value={{
        ...state,
        handleClick: handleClick,
        handleOpenModal: handleOpenModal,
        handleCloseModal: handleCloseModal,
        dispatch: dispatch,
      }}
      >
      <NavBar />
      <Container id="title">
        <Row className='justify-content-center'>
          <Col lg={12}>
            <h2>
              “You want to wake up in the morning and think the future is going
              to be great - and that’s what being a spacefaring civilization is all
              about. It’s about believing in the future and thinking that the future
              will be better than the past. And I can’t think of anything more
              exciting than going out there and being among the stars.”
            </h2>
            <h2 className="text-right">
              -Elon Musk
            </h2>
          </Col>
        </Row>
      </Container>
      <Container>
        <Row>
          {state.isError && <h3>Something is wrong ...</h3>}

          {state.loading ? <p>Loading ...</p> : <List />}

          {state.isOpen && <ModalComponent />}
        </Row>
      </Container>
    </Context.Provider>
  )
};

export default App;
export { Context, dataReducer };
