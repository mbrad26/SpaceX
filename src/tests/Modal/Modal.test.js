import React from 'react';
import { render, screen } from '@testing-library/react';
import { itemOne } from '../fixtures';
import { Context } from '../../components/App/App.js';
import ModalComponent from '../../components/Modal/Modal';

describe('Stats', () => {
  it('renders all the props', () => {
    const context = {
      isOpen: true,
      activeItem: itemOne,
      handleCloseModal: jest.fn(),
    }
    const { container } = render(
      <Context.Provider value={context}>
        <ModalComponent />
      </Context.Provider>
    );

    expect(container.firstChild).toMatchSnapshot();
  });
});
