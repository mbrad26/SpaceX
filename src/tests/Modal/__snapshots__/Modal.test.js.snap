// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stats renders all the props 1`] = `null`;
